### DISCUSSION TITLE
[eframe] That run_native is not the last command executed

### DISCUSSION THREAD

**Discussion by ricglz** [emilk/egui#1604] [Q&A] [answered]
Hi,

I'm developing a programming language for a class. In which the user can ask to plot a scatter plot or a histogram, just like you would do in Python with matplotlib. The problem is that when is attempted to run both the `plot` and `histogram` commands due the nature of `run_native` to be the last command, the second command will not be executed. Is there a way to handle this scenario?

**Comment by willser** [comment, marked as answer]
You could use [`async`](https://github.com/emilk/egui#how-do-i-use-egui-with-async).I don't know much about `plot` or `histogram`,but I think it's a good way,you can run them asynchronously at a same time with `run_native`.

In my project,I use [poll_promise::Promise](https://docs.rs/poll-promise) to handle http request asynchronously.There is an example with `channel` in [mCubed](https://github.com/4JX/mCubed/blob/61034fd7b0e2ec12b7fb96f8a894e59893df1343/main/src/ui/mod.rs#L49-L50).Thanks @4JX ❤️.

**Reply by ricglz** [reply]
Hi, attempted the [poll_promise::Promise](https://docs.rs/poll-promise) approach (as it seems like the simples). But it crashed, because what I'm trying to do is that what my CLI application does is in the main thread. While the GUI stuff, runs in the background. Basically, the opposite case of what you and [mCubed](https://github.com/4JX/mCubed/blob/61034fd7b0e2ec12b7fb96f8a894e59893df1343/main/src/ui/mod.rs#L49-L50) does.

**Reply by willser** [reply]
Hi @ricglz .I'm a little confused.Do you mean you need create a cli and a gui window at same time,and control gui by cli?If it's true,
I think integrated cli into egui window is a solution .I searched `plot-cli` and maybe you could add a `single_textedit` and button under main window,command could be entered into `textedit`
and use `std::process::Command` run it asynchronously after `button` be clicked .And after work done,show image in main window.
If I missed or confused something,could you please provide your repository url if it's public?I will look at it after work.

**Reply by ricglz** [reply]
Yeah, you got it right, I want to create a CLI that controls a GUI window. What I want is that when the user closes the window, that the normal execution of the cli continues. Just like it would happen in python with libraries such as matplotlib.

The project is [this one](https://github.com/ricglz/raoul/tree/eight-delivery) and the specific lines were I use egui [are these ones](https://github.com/ricglz/raoul/blob/eight-delivery/src/vm/mod.rs#L438-L487). Now that you mentioned `std::process::Command` maybe if I cannot end up using egui. What I could do is use `std::process::Command` to run a python script that actually does this and wait for the execution to end to continue with the operation.

**Reply by willser** [reply]
I give up :tired_face: .I tried `thread::spawn` to run eframe in another thread.But I got panic.Seems eframe must run in main thread.Sorry for no helping.Maybe other people will give you a solution.

**Reply by coderedart** [reply]
due to platform specific requirements the window event loop MUST run from mainthread. this is a hard limitation which affects winit (and eframe via extension).

I recommend that you try either spawning `Command` which will run until completion, or simply go with a different architecture like launching a service which runs in the background and is controlled by a rpc / socket. then, use the gui app to connect to the service using the rpc / socket to send it work or receive results etc..

**Reply by ricglz** [reply]
Ok, I'll think about the possible solution to this problem. Thanks a lot for your responses!
